File: src/dumps.ts
import { access, constants, mkdir, readdir, rm, stat, writeFile } from "node:fs/promises";
import { basename, join, resolve } from "node:path";
import type { Engine } from "./config.ts";
import { decryptBytes, encryptBytes } from "./crypto.ts";

export function resolveDumpsRoot(dumpDirectory: string): string {
  const abs = resolve(dumpDirectory);
  if (basename(abs).toLowerCase() === "dumps") return abs;
  return join(abs, "dumps");
}

export async function ensureDumpsRootWritable(dumpsRoot: string): Promise<void> {
  await mkdir(dumpsRoot, { recursive: true });
  try {
    await access(dumpsRoot, constants.W_OK);
  } catch {
    throw new Error(`Dumps directory is not writable: ${dumpsRoot}`);
  }
  const probe = join(dumpsRoot, `.dumpmgr-write-test-${Date.now()}`);
  try {
    await writeFile(probe, "ok");
    await rm(probe);
  } catch {
    throw new Error(`Dumps directory is not writable: ${dumpsRoot}`);
  }
}

export function dumpExtension(engine: Engine): string {
  return engine === "postgres" ? ".dump" : ".sql";
}

/** UTC timestamp: yyyy-MM-dd_HH-mm-ss */
export function dumpTimestamp(d = new Date()): string {
  const p = (n: number) => String(n).padStart(2, "0");
  return `${d.getUTCFullYear()}-${p(d.getUTCMonth() + 1)}-${p(d.getUTCDate())}_${p(d.getUTCHours())}-${p(d.getUTCMinutes())}-${p(d.getUTCSeconds())}`;
}

export function dbDumpDir(
  dumpsRoot: string,
  engine: Engine,
  itemKey: string,
): string {
  return join(dumpsRoot, engine, ...itemKey.split(":"));
}

export function dumpFileKey(itemKey: string): string {
  return itemKey.replaceAll(":", "__");
}

const ENC_ID_RE = /_enc_([A-F0-9]+)(\.(dump|sql))$/i;

export function newDumpFileName(
  engine: Engine,
  itemKey: string,
): string {
  return `${engine}_${dumpFileKey(itemKey)}_${dumpTimestamp()}${dumpExtension(engine)}`;
}

export function isEncryptedDumpName(fileName: string): boolean {
  return (
    fileName.endsWith(".enc") ||
    ENC_ID_RE.test(fileName) ||
    /_encrypted\.(dump|sql)$/i.test(fileName)
  );
}

export function dumpEncIdFromName(fileName: string): string | null {
  const m = fileName.match(ENC_ID_RE);
  return m ? m[1]!.toUpperCase() : null;
}

/** Derive a plaintext temp basename from an encrypted dump filename. */
export function plainTempNameFromEncrypted(fileName: string): string {
  if (fileName.endsWith(".enc")) return fileName.replace(/\.enc$/, "");
  return fileName
    .replace(ENC_ID_RE, "$2")
    .replace(/_encrypted(\.(dump|sql))$/i, "$1");
}

export function encryptedPathFromPlain(path: string, encId: string): string {
  if (isEncryptedDumpName(basename(path))) return path;
  return path.replace(/(\.(dump|sql))$/i, `_enc_${encId}$1`);
}

export async function listDumpFiles(
  dir: string,
  encryptedOnly: boolean,
): Promise<string[]> {
  try {
    const entries = await readdir(dir);
    return entries
      .filter((f) => {
        const isDump =
          f.endsWith(".dump") ||
          f.endsWith(".sql") ||
          f.endsWith(".dump.enc") ||
          f.endsWith(".sql.enc");
        if (!isDump) return false;
        return encryptedOnly
          ? isEncryptedDumpName(f)
          : !isEncryptedDumpName(f);
      })
      .sort()
      .reverse();
  } catch {
    return [];
  }
}

export type DumpBrowserEntry =
  | { kind: "dir"; name: string }
  | { kind: "file"; name: string };

export async function listDumpBrowserEntries(
  dir: string,
  encryptedOnly: boolean,
): Promise<DumpBrowserEntry[]> {
  let names: string[];
  try {
    names = await readdir(dir);
  } catch {
    return [];
  }
  const dirs: DumpBrowserEntry[] = [];
  const files: DumpBrowserEntry[] = [];
  for (const name of names) {
    if (name.startsWith(".")) continue;
    const full = join(dir, name);
    const st = await stat(full);
    if (st.isDirectory()) {
      dirs.push({ kind: "dir", name });
      continue;
    }
    const isDump =
      name.endsWith(".dump") ||
      name.endsWith(".sql") ||
      name.endsWith(".dump.enc") ||
      name.endsWith(".sql.enc");
    if (!isDump) continue;
    const ok = encryptedOnly
      ? isEncryptedDumpName(name)
      : !isEncryptedDumpName(name);
    if (ok) files.push({ kind: "file", name });
  }
  dirs.sort((a, b) => a.name.localeCompare(b.name));
  files.sort((a, b) => b.name.localeCompare(a.name));
  return [...dirs, ...files];
}

export async function encryptDumpFile(
  path: string,
  key: CryptoKey,
  encId: string,
): Promise<string> {
  const raw = new Uint8Array(await Bun.file(path).arrayBuffer());
  const enc = await encryptBytes(key, raw);
  const outPath = encryptedPathFromPlain(path, encId);
  await Bun.write(outPath, enc);
  if (outPath !== path) await rm(path);
  return outPath;
}

export async function decryptDumpToTemp(
  encPath: string,
  key: CryptoKey,
  tempPath: string,
): Promise<void> {
  const data = new Uint8Array(await Bun.file(encPath).arrayBuffer());
  const plain = await decryptBytes(key, data);
  await Bun.write(tempPath, plain);
}

export async function hasEncryptedDumpsWithEncId(
  dumpsRoot: string,
  encId: string,
): Promise<boolean> {
  const want = encId.toUpperCase();
  const walk = async (dir: string): Promise<boolean> => {
    let entries: string[];
    try {
      entries = await readdir(dir);
    } catch {
      return false;
    }
    for (const name of entries) {
      const full = join(dir, name);
      const st = await stat(full);
      if (st.isDirectory()) {
        if (await walk(full)) return true;
        continue;
      }
      if (dumpEncIdFromName(name) === want) return true;
    }
    return false;
  };
  return walk(dumpsRoot);
}

export async function reencryptAllDumps(
  dumpsRoot: string,
  oldKey: CryptoKey,
  newKey: CryptoKey,
  encId?: string,
): Promise<number> {
  let count = 0;
  const want = encId?.toUpperCase();
  const walk = async (dir: string) => {
    let entries: string[];
    try {
      entries = await readdir(dir);
    } catch {
      return;
    }
    for (const name of entries) {
      const full = join(dir, name);
      const st = await stat(full);
      if (st.isDirectory()) {
        await walk(full);
        continue;
      }
      if (!isEncryptedDumpName(name)) continue;
      if (want) {
        const id = dumpEncIdFromName(name);
        if (id !== want) continue;
      }
      const data = new Uint8Array(await Bun.file(full).arrayBuffer());
      const plain = await decryptBytes(oldKey, data);
      const enc = await encryptBytes(newKey, plain);
      await Bun.write(full, enc);
      count++;
    }
  };
  await walk(dumpsRoot);
  return count;
}

export async function deleteEncryptedDumpsWithEncId(
  dumpsRoot: string,
  encId: string,
): Promise<number> {
  let count = 0;
  const want = encId.toUpperCase();
  const walk = async (dir: string) => {
    let entries: string[];
    try {
      entries = await readdir(dir);
    } catch {
      return;
    }
    for (const name of entries) {
      const full = join(dir, name);
      const st = await stat(full);
      if (st.isDirectory()) {
        await walk(full);
        continue;
      }
      if (dumpEncIdFromName(name) === want) {
        await rm(full, { force: true });
        count++;
      }
    }
  };
  await walk(dumpsRoot);
  return count;
}

export async function deleteAllDumps(dumpsRoot: string): Promise<void> {
  await rm(dumpsRoot, { recursive: true, force: true });
  await mkdir(dumpsRoot, { recursive: true });
}

export function formatBytes(n: number): string {
  if (n < 1024) return `${n} B`;
  if (n < 1024 ** 2) return `${(n / 1024).toFixed(1)} KB`;
  if (n < 1024 ** 3) return `${(n / 1024 ** 2).toFixed(1)} MB`;
  return `${(n / 1024 ** 3).toFixed(2)} GB`;
}

export function formatDuration(ms: number): string {
  return `${(ms / 1000).toFixed(1)}s`;
}

File: src/internal/dumps/dumps.go
package dumps

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/md-redwan-hossain/dumpmgr/src/internal/crypto"
)

var encIDRE = regexp.MustCompile(`(?i)_enc_([A-F0-9]+)(\.dump)$`)

func ResolveRoot(dumpDirectory string) string {
	abs, err := filepath.Abs(dumpDirectory)
	if err != nil {
		abs = dumpDirectory
	}
	if strings.EqualFold(filepath.Base(abs), "dumps") {
		return abs
	}
	return filepath.Join(abs, "dumps")
}

func EnsureRootWritable(dumpsRoot string) error {
	if err := os.MkdirAll(dumpsRoot, 0o755); err != nil {
		return err
	}
	probe := filepath.Join(dumpsRoot, fmt.Sprintf(".dumpmgr-write-test-%d", time.Now().UnixNano()))
	if err := os.WriteFile(probe, []byte("ok"), 0o644); err != nil {
		return fmt.Errorf("dumps directory is not writable: %s", dumpsRoot)
	}
	return os.Remove(probe)
}

func DumpExtension() string {
	return ".dump"
}

func DumpTimestamp(t time.Time) string {
	return fmt.Sprintf("%04d-%02d-%02d_%02d-%02d-%02d",
		t.UTC().Year(), t.UTC().Month(), t.UTC().Day(),
		t.UTC().Hour(), t.UTC().Minute(), t.UTC().Second())
}

func DBDumpDir(dumpsRoot, itemKey string) string {
	parts := strings.Split(itemKey, ":")
	return filepath.Join(append([]string{dumpsRoot}, parts...)...)
}

func DumpFileKey(itemKey string) string {
	return strings.ReplaceAll(itemKey, ":", "__")
}

func NewDumpFileName(itemKey string) string {
	return fmt.Sprintf("%s_%s%s", DumpFileKey(itemKey), DumpTimestamp(time.Now()), DumpExtension())
}

func IsEncryptedDumpName(fileName string) bool {
	if strings.HasSuffix(fileName, ".enc") {
		return true
	}
	if encIDRE.MatchString(fileName) {
		return true
	}
	return strings.HasSuffix(strings.ToLower(fileName), "_encrypted.dump")
}

func DumpEncIDFromName(fileName string) string {
	m := encIDRE.FindStringSubmatch(fileName)
	if len(m) < 2 {
		return ""
	}
	return strings.ToUpper(m[1])
}

func PlainTempNameFromEncrypted(fileName string) string {
	if strings.HasSuffix(fileName, ".enc") {
		return strings.TrimSuffix(fileName, ".enc")
	}
	name := encIDRE.ReplaceAllString(fileName, "$2")
	re := regexp.MustCompile(`(?i)_encrypted(\.dump)$`)
	return re.ReplaceAllString(name, "$1")
}

func EncryptedPathFromPlain(path, encID string) string {
	base := filepath.Base(path)
	if IsEncryptedDumpName(base) {
		return path
	}
	re := regexp.MustCompile(`(?i)(\.dump)$`)
	return re.ReplaceAllString(path, fmt.Sprintf("_enc_%s$1", encID))
}

func ListDumpFiles(dir string, encryptedOnly bool) ([]string, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, nil
	}
	var files []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		name := e.Name()
		isDump := strings.HasSuffix(name, ".dump") || strings.HasSuffix(name, ".dump.enc")
		if !isDump {
			continue
		}
		enc := IsEncryptedDumpName(name)
		if encryptedOnly && !enc {
			continue
		}
		if !encryptedOnly && enc {
			continue
		}
		files = append(files, name)
	}
	// sort descending
	for i := 0; i < len(files); i++ {
		for j := i + 1; j < len(files); j++ {
			if files[j] > files[i] {
				files[i], files[j] = files[j], files[i]
			}
		}
	}
	return files, nil
}

type BrowserEntry struct {
	Kind BrowserEntryKind
	Name string
}

type BrowserEntryKind int

const (
	EntryDir BrowserEntryKind = iota
	EntryFile
)

func ListBrowserEntries(dir string, encryptedOnly bool) ([]BrowserEntry, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, nil
	}
	var dirs, files []BrowserEntry
	for _, e := range entries {
		name := e.Name()
		if strings.HasPrefix(name, ".") {
			continue
		}
		info, err := e.Info()
		if err != nil {
			continue
		}
		if info.IsDir() {
			dirs = append(dirs, BrowserEntry{Kind: EntryDir, Name: name})
			continue
		}
		isDump := strings.HasSuffix(name, ".dump") || strings.HasSuffix(name, ".dump.enc")
		if !isDump {
			continue
		}
		enc := IsEncryptedDumpName(name)
		if encryptedOnly && !enc {
			continue
		}
		if !encryptedOnly && enc {
			continue
		}
		files = append(files, BrowserEntry{Kind: EntryFile, Name: name})
	}
	sortEntries(dirs)
	sortEntriesDesc(files)
	return append(dirs, files...), nil
}

func sortEntries(entries []BrowserEntry) {
	for i := 0; i < len(entries); i++ {
		for j := i + 1; j < len(entries); j++ {
			if entries[j].Name < entries[i].Name {
				entries[i], entries[j] = entries[j], entries[i]
			}
		}
	}
}

func sortEntriesDesc(entries []BrowserEntry) {
	for i := 0; i < len(entries); i++ {
		for j := i + 1; j < len(entries); j++ {
			if entries[j].Name > entries[i].Name {
				entries[i], entries[j] = entries[j], entries[i]
			}
		}
	}
}

func EncryptDumpFile(path string, key []byte, encID string) (string, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	enc, err := crypto.EncryptBytes(key, raw)
	if err != nil {
		return "", err
	}
	outPath := EncryptedPathFromPlain(path, encID)
	if err := os.WriteFile(outPath, enc, 0o644); err != nil {
		return "", err
	}
	if outPath != path {
		_ = os.Remove(path)
	}
	return outPath, nil
}

func DecryptDumpToTemp(encPath string, key []byte, tempPath string) error {
	data, err := os.ReadFile(encPath)
	if err != nil {
		return err
	}
	plain, err := crypto.DecryptBytes(key, data)
	if err != nil {
		return err
	}
	return os.WriteFile(tempPath, plain, 0o600)
}

func HasEncryptedDumpsWithEncID(dumpsRoot, encID string) (bool, error) {
	want := strings.ToUpper(encID)
	var walk func(string) (bool, error)
	walk = func(dir string) (bool, error) {
		entries, err := os.ReadDir(dir)
		if err != nil {
			return false, nil
		}
		for _, e := range entries {
			full := filepath.Join(dir, e.Name())
			if e.IsDir() {
				found, err := walk(full)
				if err != nil || found {
					return found, err
				}
				continue
			}
			if DumpEncIDFromName(e.Name()) == want {
				return true, nil
			}
		}
		return false, nil
	}
	return walk(dumpsRoot)
}

func ReencryptAllDumps(dumpsRoot string, oldKey, newKey []byte, encID string) (int, error) {
	want := strings.ToUpper(encID)
	count := 0
	var walk func(string) error
	walk = func(dir string) error {
		entries, err := os.ReadDir(dir)
		if err != nil {
			return nil
		}
		for _, e := range entries {
			full := filepath.Join(dir, e.Name())
			if e.IsDir() {
				if err := walk(full); err != nil {
					return err
				}
				continue
			}
			if !IsEncryptedDumpName(e.Name()) {
				continue
			}
			if want != "" && DumpEncIDFromName(e.Name()) != want {
				continue
			}
			data, err := os.ReadFile(full)
			if err != nil {
				return err
			}
			plain, err := crypto.DecryptBytes(oldKey, data)
			if err != nil {
				return err
			}
			enc, err := crypto.EncryptBytes(newKey, plain)
			if err != nil {
				return err
			}
			if err := os.WriteFile(full, enc, 0o644); err != nil {
				return err
			}
			count++
		}
		return nil
	}
	if err := walk(dumpsRoot); err != nil {
		return 0, err
	}
	return count, nil
}

func DeleteEncryptedDumpsWithEncID(dumpsRoot, encID string) (int, error) {
	want := strings.ToUpper(encID)
	count := 0
	var walk func(string) error
	walk = func(dir string) error {
		entries, err := os.ReadDir(dir)
		if err != nil {
			return nil
		}
		for _, e := range entries {
			full := filepath.Join(dir, e.Name())
			if e.IsDir() {
				if err := walk(full); err != nil {
					return err
				}
				continue
			}
			if DumpEncIDFromName(e.Name()) == want {
				if err := os.Remove(full); err != nil {
					return err
				}
				count++
			}
		}
		return nil
	}
	if err := walk(dumpsRoot); err != nil {
		return 0, err
	}
	return count, nil
}

func FormatBytes(n int64) string {
	if n < 1024 {
		return fmt.Sprintf("%d B", n)
	}
	if n < 1024*1024 {
		return fmt.Sprintf("%.1f KB", float64(n)/1024)
	}
	if n < 1024*1024*1024 {
		return fmt.Sprintf("%.1f MB", float64(n)/(1024*1024))
	}
	return fmt.Sprintf("%.2f GB", float64(n)/(1024*1024*1024))
}

func FormatDuration(ms float64) string {
	return fmt.Sprintf("%.1fs", math.Round(ms/100)/10)
}

func FileSize(path string) (int64, error) {
	info, err := os.Stat(path)
	if err != nil {
		return 0, err
	}
	return info.Size(), nil
}
